### DISCUSSION TITLE
Get all atoms on all paths between two atoms in a molecule in C++?

### DISCUSSION THREAD

**Discussion by NiklasKappel** [rdkit/rdkit#5221] [Q&A] [answered]
As an example, say I have an [OPD](https://en.wikipedia.org/wiki/O-Phenylenediamine) molecule represented by a `RDKit::ROMol`. There are two "paths" connecting the two Nitrogen atoms in the molecule: A long path around the ring N1 - C1 - C2 - C3 - C4 - C5 - C6 - N2 and a short direct path N1 - C1 - C6 - N2.

Say I have the atom indexes `idx1` and `idx2`. Is there an easy way to define a function `std::vector<std::vector<int>> get_paths(int idx1, int idx2, RDKit::ROMol mol)` that returns a list of paths between the atoms at `idx1` and `idx2`, where each path is given by a list of indexes of the atoms along the path (i.e. one list of 8 and one list of 4 indexes in the above example)?

Right now I am looking into `RDKit::ROMol.getTopology()` returning `boost::adjacency_list` to see if there is an algorithm for that.

**Comment by bp-kelley** [comment, marked as answer]
You'll need to implement a simple DFS (depth first traversal) to do this, I don't think one exists in the code base, but it's pretty easy to implement.  Here is a bit of Python code that might help you on your way:

```
from rdkit import Chem

# recursive depth first search
#  when you hit the target atom, record the path and return
#  don't loop back on atoms in the path
def _dfs(atom, end_idx, path, paths):
    for bond in atom.GetBonds():
        oatom = bond.GetOtherAtom(atom)
        idx = oatom.GetIdx()

        if idx == end_idx:
            # end goal
            paths.append(list(path) + [idx])
            return
        
        elif idx in path:
            # nope, we visited this atom
            continue
        
        path.append(idx)
        _dfs(oatom, end_idx, path, paths)
        path.pop()


def dfs(mol, start_idx, end_idx):
    paths = []
    atom = mol.GetAtomWithIdx(start_idx)
    for bond in atom.GetBonds():
        oatom = bond.GetOtherAtom(atom)
        idx = oatom.GetIdx()
        path = [start_idx, idx]
        _dfs(oatom, end_idx, path, paths)

    return paths

m = Chem.MolFromSmiles("Nc1ccccc1N")
print(dfs(m, 0, 7))
#  [[0, 1, 2, 3, 4, 5, 6, 7], [0, 1, 6, 7]]
```

**Reply by raymondlanger** [reply]
The code is very helpful, but I think if `start_idx` and `end_idx` are equal, the returned list should only contain this one value. This currently not the case.

**Reply by greglandrum** [reply]
I missed this one at the time, but there's a function `Chem.GetShortestPath()` that returns the shortest atom path between two atoms in a molecule.
